program/src/processor.rs

```diff
@@ -669,15 +669,57 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
 mod tests {
     use {super::*, proptest::prelude::*};
 
+    const BENCH_TOKEN_SUPPLY: u64 = 1_000_000_000 * 1_000_000_000; // 1 billion with 9 decimals
+
+    #[test]
+    fn minimum_rewards_per_token() {
+        // 1 lamport (arithmetic minimum)
+        let minimum_reward = 1;
+        let result = calculate_rewards_per_token(minimum_reward, BENCH_TOKEN_SUPPLY).unwrap();
+        assert_ne!(result, 0);
+
+        // Anything below the minimum should return zero.
+        let result = calculate_rewards_per_token(minimum_reward - 1, BENCH_TOKEN_SUPPLY).unwrap();
+        assert_eq!(result, 0);
+    }
+
+    #[test]
+    fn maximum_rewards_per_token() {
+        // u64::MAX (not really practical, but shows that we're ok)
+        let maximum_reward = u64::MAX;
+        let _ = calculate_rewards_per_token(maximum_reward, BENCH_TOKEN_SUPPLY).unwrap();
+    }
+
+    #[test]
+    fn minimum_eligible_rewards() {
+        // 1 / 1e18 lamports per token
+        let minimum_marginal_rewards_per_token = 1;
+        let result = calculate_eligible_rewards(
+            minimum_marginal_rewards_per_token,
+            0,
+            BENCH_TOKEN_SUPPLY, // 100% of the supply.
+        )
+        .unwrap();
+        assert_ne!(result, 0);
+
+        // Anything below the minimum should return zero.
+        let result = calculate_eligible_rewards(
+            minimum_marginal_rewards_per_token - 1,
+            0,
+            BENCH_TOKEN_SUPPLY, // 100% of the supply.
+        )
+        .unwrap();
+        assert_eq!(result, 0);
+    }
```

Comment:
This test is great -- can you add about another for minimum eligible rewards for someone with 1 PAL (ie 1e9 tokens)? That way we'll know how often an "average" user can harvest (which I think should come out to every 1 SOL?). It can be done in a separate PR if you prefer

Reply:
Sure thing, added!

With a scaling factor of `1e18`, the minimum marginal rate is actually the same as the minimum system-level reward a holder could harvest against.

```
rewards_per_token = (rewards * scaling) / supply
rewards_per_token = (rewards * 1e18) / 1e18
rewards_per_token = rewards
```

So if you have 1 PAL you can harvest every `1 / e9` lamports (reward) and if you have .000_000_001 PAL you can harvest every 1 lamport.

Reply:
I guess both are technically 1 lamport though, since you can't distribute less than `1u64`.